src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -62,6 +64,287 @@ static ShardPlacement * TupleToShardPlacement(TupleDesc tupleDesc,
+		char *shardQualifiedName = NULL;
+		char *quotedShardName = NULL;
+		AppendShardIdToName(&shardName, shardId);
+
+		shardQualifiedName = quote_qualified_identifier(schemaName, shardName);
+		quotedShardName = quote_literal_cstr(shardQualifiedName);
+
+		if (isFirstShard)
+		{
+			appendStringInfo(selectQuery, "Select ");
+			isFirstShard = !isFirstShard;
+		}
+
+		if (sizeFunctionType == CITUS_TOTAL_RELATION_SIZE)
+		{
+			appendStringInfo(selectQuery, INNER_SHARD_TOTAL_RELATION_SIZE_QUERY,
+							 quotedShardName);
+		}
+		else if (sizeFunctionType == CITUS_TABLE_SIZE)
+		{
+			appendStringInfo(selectQuery, INNER_SHARD_TABLE_SIZE_QUERY,
+							 quotedShardName);
+		}
+		else if (sizeFunctionType == CITUS_RELATION_SIZE)
+		{
+			appendStringInfo(selectQuery, INNER_SHARD_RELATION_SIZE_QUERY,
+							 quotedShardName);
+		}
+	}
+
+	/* there is no more size function so just add 0 instead of it */
+	appendStringInfo(selectQuery, "0;");
+	return selectQuery;
+}
+
+
+/*
+ * SupportFunctionForCitusSize is a helper function for all citus size functions. It
+ * first checks whether the table is distributed and size query can be run on it. For
+ * distributed tables with streaming replications connection to each node has to be
+ * established. For reference tables, connecting to one of the node is enough.
+ */
+uint64
+SupportFunctionForCitusSize(Oid distributedRelationId, CitusSizeFunctionTypes
+							sizeFunctionType)
+{
+	DistTableCacheEntry *distributedTableCacheEntry = NULL;
+
+	List *workerNodeList = WorkerNodeList();
+	ListCell *workerNodeCell = NULL;
+
+	uint64 totalRelationSize = 0;
+
+	MultiConnection *connection = NULL;
+
+	distributedTableCacheEntry = DistributedTableCacheEntry(distributedRelationId);
+
+	foreach(workerNodeCell, workerNodeList)
+	{
+		PGresult *queryResult = NULL;
+		StringInfo tableSizeQuery = NULL;
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		char *workerNodeName = workerNode->workerName;
+		uint32 workerNodePort = workerNode->workerPort;
+		char *tableSizeString;
+		int executeCommand = 0;
+		uint32 connectionFlag = 0;
+
+		List *intervalsOnNode = GetShardIntervalsOnNode(workerNode,
+														distributedRelationId);
+
+
+		tableSizeQuery = GenerateSizeQueryOnMultiplePlacements(distributedRelationId,
+															   intervalsOnNode,
+															   sizeFunctionType);
+
+		connection = GetNodeConnection(connectionFlag, workerNodeName, workerNodePort);
+
+		executeCommand = ExecuteOptionalRemoteCommand(connection, tableSizeQuery->data,
+													  &queryResult);
+
+		/*
+		 * If there is a connection error and the table is not reference table,
+		 * close the connection and quit. If it is a reference table continue
+		 * to get the size from another node.
+		 */
+		if (executeCommand != 0)
+		{
+			PQclear(queryResult);
+
+			ForgetResults(connection);
+
+			if (distributedTableCacheEntry->partitionMethod == DISTRIBUTE_BY_NONE)
```

Comment:
Contrary to previous  function, we can use PartitionMethod(relationId) here to avoid using distributedTableCacheEntry.